fix: prefer to set mod time to oldes allowed by zip

shouze opened the pull request:
# What does it fix?

zip files generated by this plugin are great because if we produce a zip file of the same files in two distinct points in time it will produce a zip archive with exactly the same resulting hash :heart:.

But... in go `time.Time{}` is the zero time, the [documenation if very clear on it](https://golang.org/pkg/time/#Time):
> The zero value of type Time is January 1, year 1, 00:00:00.000000000 UTC

This leads to 2 major issues:
1. For many zip unarchivers (see note below), the oldest supported time is `1980-01-01 0:00 UTC` so we potentialy generate archives not supported by any zip unarchiver :zipper_mouth_face:.
2. On my Linux system at least, the files once unarchive on my local file system have dates in 2049, so in the future. And I can tell you that it leads some of my AWS lambda to crash when I distribute them through a zip created but this plugin :boom:.


# Notes

As spotted in https://www.mindprod.com/jgloss/zip.html:
> The format does not support dates prior to 1980-01-01 0:00 UTC . Avoid file dates 1980-01-01 or earlier (local or UTC time).

I'm not sur to generate the given date the best way that can be done in go, by the way I've tested the compiled module on a real situation and it fixes the issue I'm facing :wink:.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-provider-archive?pullRequest=93)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-provider-archive?pullRequest=93)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-provider-archive?pullRequest=93).</sub>

shouze commented:
ping @appilon or @kmoe would love your feedback :100:

kmoe reviewed (changes requested):
Thanks - this looks like a very sensible fix!

[`zip#FileHeader.SetModTime()`](https://golang.org/pkg/archive/zip/#FileHeader.SetModTime) is marked as deprecated in the stdlib docs in favour of setting `FileHeader.Modified` directly. Judging by recent changes to archive/zip I think it's worth switching to `FileHeader.Modified` so we can remove the "zero value" comment added in https://github.com/hashicorp/terraform-provider-archive/pull/16.

Please also add an acceptance test that checks the modified time of the archived file. You could add this to `TestZipArchiver_FileModified` or make a new test.

shouze commented:
> Thanks - this looks like a very sensible fix!
> 
> [`zip#FileHeader.SetModTime()`](https://golang.org/pkg/archive/zip/#FileHeader.SetModTime) is marked as deprecated in the stdlib docs in favour of setting `FileHeader.Modified` directly. Judging by recent changes to archive/zip I think it's worth switching to `FileHeader.Modified` so we can remove the "fh.Modified alone isn't enough when using a zero value" comment added in #16.
> 
> Please also add an acceptance test that checks the modified time of the archived file. You could add this to `TestZipArchiver_FileModified` or make a new test.

@kmoe I've made all requested changes, LGTM?

Dezzj reviewed (commented):
Veiwed
